main/src/ca/uwaterloo/flix/language/phase/optimizer/Inliner.scala

```diff
@@ -371,6 +373,195 @@ object Inliner {
   }
 
+  /**
+    * Evaluate match if possible.
+    *
+    * The match rules are iterated top to bottom, at each step either:
+    *   - Removing the rule if it can never match (either the patten cannot match or the guard is guaranteed false)
+    *   - Removing the match if a rule is known to match (either `Some(12)` matching `Some(x)` or a lenient pattern like `_`)
+    *   - Returning the rules as is if the rule cannot be determined statically.
+    */
+  @tailrec
+  private def reduceMatch(exp: MonoAst.Expr, rules: List[MonoAst.MatchRule], tpe: Type, eff: Type, loc: SourceLocation)(implicit sym0: Symbol.DefnSym, sctx: SharedContext): Expr = {
+    rules match {
+      case MonoAst.MatchRule(pat, guardOpt, ruleExp) :: rest =>
+        matchRule(exp, pat, guardOpt) match {
+          case MatchResult.Match(binders) =>
+            // Guaranteed rule - convert to let binders.
+            sctx.changed.putIfAbsent(sym0, ())
+            bindPatterns(binders.toSeq, ruleExp, loc)
+          case MatchResult.NoMatch =>
+            // Impossible rule - delete and continue.
+            sctx.changed.putIfAbsent(sym0, ())
+            reduceMatch(exp, rest, tpe, eff, loc)
+          case MatchResult.Unknown =>
+            // Match is unknown - do nothing.
+            Expr.Match(exp, rules, tpe, eff, loc)
+        }
+      case Nil =>
+        // Do nothing.
+        Expr.Match(exp, rules, tpe, eff, loc)
+    }
+  }
+
+  private sealed trait FuzzyBool
+
+  private object FuzzyBool {
+
+    case object True extends FuzzyBool
+
+    case object False extends FuzzyBool
+
+    case object Unknown extends FuzzyBool
+
+  }
+
+  /** Returns the fuzzy evaluation of `exp` as a boolean. */
+  private def evalBoolExpression(exp: MonoAst.Expr): FuzzyBool = exp match {
+    case Expr.Cst(Constant.Bool(true), _, _) => FuzzyBool.True
+    case Expr.Cst(Constant.Bool(false), _, _) => FuzzyBool.False
+    case _ => FuzzyBool.Unknown
+  }
+
+  private sealed trait MatchResult
+
+  private object MatchResult {
+
+    case class Match(binders: Chain[(Option[Pattern.Var], MonoAst.Expr)]) extends MatchResult
+
+    case object NoMatch extends MatchResult
+
+    case object Unknown extends MatchResult
+
+    def emptyMatch(): MatchResult =
+      Match(Chain.empty)
+
+    def singleMatch(pat: Option[Pattern.Var], exp: MonoAst.Expr): MatchResult =
+      Match(Chain((pat, exp)))
+
+    def matchFromBool(b: Boolean): MatchResult =
+      if (b) emptyMatch() else NoMatch
+
+    /**
+      * Concatenates two match results.
+      *
+      * If either result is [[NoMatch]], the output is [[NoMatch]].
+      *
+      * Then if either result is [[Unknown]], the output is [[Unknown]].
+      *
+      * Then (both results now being [[Match]]) the binders are concatenated.
+      */
+    def concat(mr1: MatchResult, mr2: MatchResult): MatchResult = (mr1, mr2) match {
+      case (MatchResult.NoMatch, _) | (_, MatchResult.NoMatch) =>
```

Comment:
We dont use or patterns anywhere else in the compiler; let us not start here.

Reply:
https://github.com/flix/flix/blob/4df8a6dd7f28273951977de25bae073a2a02f2c4/main/src/ca/uwaterloo/flix/language/ast/TokenKind.scala#L394-L404

https://github.com/flix/flix/blob/4df8a6dd7f28273951977de25bae073a2a02f2c4/main/src/ca/uwaterloo/flix/language/ast/SimpleType.scala#L150-L153

https://github.com/flix/flix/blob/4df8a6dd7f28273951977de25bae073a2a02f2c4/main/src/ca/uwaterloo/flix/language/phase/Monomorpher.scala#L281-L283

https://github.com/flix/flix/blob/4df8a6dd7f28273951977de25bae073a2a02f2c4/main/src/ca/uwaterloo/flix/language/phase/Parser2.scala#L1614-L1630

etc etc etc